File: AmneziaAndroid/core/src/main/java/org/amnezia/core/models/ServerProfile.kt
package org.amnezia.core.models

import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class ServerProfile(
    val id: String = UUID.randomUUID().toString(),
    var name: String = "Amnezia Server",
    var protocolType: ProtocolType = ProtocolType.AMNEZIA_WG,
    var host: String = "",
    var port: Int = 51820,
    var clientIp: String = "10.0.0.2/32",
    var dnsServers: List<String> = listOf("1.1.1.1", "1.0.0.1"),
    var publicKey: String = "",
    var privateKey: String = "",
    var presharedKey: String? = null,
    var allowedIps: List<String> = listOf("0.0.0.0/0", "::/0"),
    // AmneziaWG Obfuscation parameters
    var jc: Int? = null,
    var jmin: Int? = null,
    var jmax: Int? = null,
    var s1: Int? = null,
    var s2: Int? = null,
    var h1: Long? = null,
    var h2: Long? = null,
    var h3: Long? = null,
    var h4: Long? = null,
    var mtu: Int = 1280,
    var pingMs: Long? = null,
    var flagEmoji: String = "🌐"
) {
    val endpointString: String
        get() = if (host.isNotEmpty()) "$host:$port" else "Not configured"

    val protocol: ProtocolType
        get() = protocolType
}

File: AmneziaCore/Models/ServerProfile.swift
import Foundation

public enum ProtocolType: String, Codable, Sendable, CaseIterable {
    case amneziaWg = "AmneziaWG"
    case wireGuard = "WireGuard"
    case openVpn = "OpenVPN"
    case xray = "XRay"

    public var badgeTitle: String {
        switch self {
        case .amneziaWg: return "AWG (Obfuscated)"
        case .wireGuard: return "WireGuard"
        case .openVpn: return "OpenVPN"
        case .xray: return "XRay"
        }
    }
}

/// Unified VPN Server Configuration Profile
public struct ServerProfile: Identifiable, Codable, Equatable, Sendable {
    public let id: UUID
    public var name: String
    public var protocolType: ProtocolType
    
    // Endpoint
    public var endpointHost: String
    public var endpointPort: UInt16
    
    // Cryptographic Keys
    public var clientPrivateKey: String
    public var serverPublicKey: String
    public var presharedKey: String?
    
    // Network Settings
    public var clientAddresses: [String]
    public var dnsServers: [String]
    public var allowedIps: [String]
    public var mtu: UInt16?
    public var persistentKeepalive: UInt16?
    
    // AmneziaWG Obfuscation (if applicable)
    public var awgParameters: AwgParameters?
    
    // Metadata
    public var rawConfig: String?
    public var createdAt: Date
    public var lastConnectedAt: Date?
    public var pingMs: Int?
    public var countryCode: String?

    public init(
        id: UUID = UUID(),
        name: String,
        protocolType: ProtocolType = .amneziaWg,
        endpointHost: String,
        endpointPort: UInt16 = 51820,
        clientPrivateKey: String,
        serverPublicKey: String,
        presharedKey: String? = nil,
        clientAddresses: [String],
        dnsServers: [String] = ["1.1.1.1", "1.0.0.1"],
        allowedIps: [String] = ["0.0.0.0/0", "::/0"],
        mtu: UInt16? = nil,
        persistentKeepalive: UInt16? = 25,
        awgParameters: AwgParameters? = nil,
        rawConfig: String? = nil,
        createdAt: Date = Date(),
        lastConnectedAt: Date? = nil,
        pingMs: Int? = nil,
        countryCode: String? = nil
    ) {
        self.id = id
        self.name = name
        self.protocolType = protocolType
        self.endpointHost = endpointHost
        self.endpointPort = endpointPort
        self.clientPrivateKey = clientPrivateKey
        self.serverPublicKey = serverPublicKey
        self.presharedKey = presharedKey
        self.clientAddresses = clientAddresses
        self.dnsServers = dnsServers
        self.allowedIps = allowedIps
        self.mtu = mtu
        self.persistentKeepalive = persistentKeepalive
        self.awgParameters = awgParameters
        self.rawConfig = rawConfig
        self.createdAt = createdAt
        self.lastConnectedAt = lastConnectedAt
        self.pingMs = pingMs
        self.countryCode = countryCode
    }

    /// Full formatted endpoint "host:port"
    public var endpointString: String {
        "\(endpointHost):\(endpointPort)"
    }

    /// Country emoji flag based on countryCode if available, else a globe
    public var flagEmoji: String {
        guard let countryCode = countryCode?.uppercased(), countryCode.count == 2 else {
            return "🌐"
        }
        let base: UInt32 = 127397
        var scalarView = String.UnicodeScalarView()
        for scalar in countryCode.unicodeScalars {
            if let newScalar = UnicodeScalar(base + scalar.value) {
                scalarView.append(newScalar)
            }
        }
        return String(scalarView)
    }
}
